client/web/src/components/popovers/ColorPicker.vue

```diff
@@ -115,13 +117,197 @@
 
 <script lang="ts">
 import { defineComponent } from "vue";
+import { clamp, HSV2RGB, RGB2HSV, RGB, HSV } from "../../lib/utils";
+
+interface Rect {
+	width: number;
+	height: number;
+	left: number;
+	top: number;
+}
+
+interface ColorPickerData {
+	color: HSV;
+	hue: {
+		rect: Rect;
+	};
+	saturation: {
+		rect: Rect;
+	};
+	opacity: {
+		rect: Rect;
+	};
+}
+
+type ColorPickerState = "idle" | "move_hue" | "move_opacity" | "move_saturation";
 
 export default defineComponent({
+	colorPicker: {
+		color: {
+			h: 0,
+			s: 0,
+			v: 0,
+			a: 1,
+		},
+		hue: {
+			rect: {
+				width: 0,
+				height: 0,
+				top: 0,
+				left: 0,
+			},
+		},
+		opacity: {
+			rect: {
+				width: 0,
+				height: 0,
+				top: 0,
+				left: 0,
+			},
+		},
+		saturation: {
+			rect: {
+				width: 0,
+				height: 0,
+				top: 0,
+				left: 0,
+			},
+		},
+	} as ColorPickerData,
 	components: {},
-	props: {},
+	props: {
+		color: {
+			type: Object,
+		},
+	},
+	data() {
+		return {
+			state: "idle" as ColorPickerState,
+		};
+	},
+	mounted() {
+		this.$watch("color", this.updateColor, { immediate: true });
+	},
+	unmounted() {
+		this.removeEvents();
+	},
 	methods: {
-		notImplemented() {
-			alert("Color picker is not functional yet");
+		addEvents() {
+			document.addEventListener("pointermove", this.onPointerMove);
+			document.addEventListener("pointerup", this.onPointerUp);
+		},
+
+		removeEvents() {
+			document.removeEventListener("pointermove", this.onPointerMove);
+			document.removeEventListener("pointerup", this.onPointerUp);
+		},
+
+		onPointerDown(e: PointerEvent) {
+			const target = e.currentTarget as Element;
+			const picker = target.getAttribute("data-picker-type");
+
+			if (picker && this.state === "idle") {
+				this.addEvents();
+				this.state = `move_${picker}` as ColorPickerState;
+				this.updateRects();
+				this.onPointerMove(e);
+			}
+		},
+
+		onPointerMove(e: PointerEvent) {
+			const { colorPicker } = this.$options as { colorPicker: ColorPickerData };
+
+			if (this.state === "move_hue") {
+				this.setHuePosition(e.y - colorPicker.hue.rect.top);
+			} else if (this.state === "move_opacity") {
+				this.setOpacityPosition(e.y - colorPicker.opacity.rect.top);
+			} else if (this.state === "move_saturation") {
+				this.setSaturationPosition(e.x - colorPicker.saturation.rect.left, e.y - colorPicker.saturation.rect.top);
+			}
+
+			if (this.state !== "idle") {
+				this.updateHue();
+				this.$emit("update:color", HSV2RGB(colorPicker.color));
```

Comment:
Nice! I didn't know about the `emit` feature! Perhaps after this PR you can add that approach for other widgets we use like `RadioInput.vue` which uses a weird callback-based approach.